geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java

```diff
@@ -1160,6 +1169,7 @@ private boolean acquiredPrimaryLock() {
     try {
       synchronized (this) {
         if (isHosting() && (isVolunteering() || isBecomingPrimary())) {
+          hasBecomePrimary = isBecomingPrimary();
```

Comment:
This might be better placed inside the if statement on line 1177, since we haven't actually become primary until `requestPrimaryState(IS_PRIMARY_HOSTING)` returns. Also, once this is set to true, there is no way for it to be set to false, which means that if a bucket becomes primary, then stops being primary (due to a rebalance, for example) this field will still be true. It might be better to instead change the visibility of the `isBecomingPrimary()` method to allow it to be called from `BucketRegionQueue.beforeAcquiringPrimaryState()`.

Reply:
This needs to be set before beforeAcquiringPrimaryState is called. It indicates that bucket is recovering after restart to become primary.

Reply:
Thanks for comments. For rebalance and recovery, same behavior is expected. Added reseting of flag.

Reply:
I think there is a timing window where `isVoluteering()` returns true in the if statement above, but `isBecomingPrimary()` does not. After the call to `setPrimaryMember()` on line 1182, this bucket will be the primary, but `hasBecomePrimary` will not have been set to true.

I feel that a better solution might be to have a method in `BucketAdvisor`:
```
  boolean isPrimaryOrVolunteering() {
    return isVolunteering() || isBecomingPrimary() || isPrimary();
  }
```
and then call that method in `BucketRegionQueue`:
```
        BucketAdvisor parent = getParentAdvisor(getBucketAdvisor());
        if (parent.isPrimaryOrVolunteering()) {
          return;
        }
```

Reply:
But, what we need is just indication that we are becoming primary (not volunteering). This is the reason, we are checking only isBecomingPrimary(). Logic is, if bucket is becoming primary (after restart or rebalance), we do not mark all events as possible duplicates (since we will get image from previous primary).

Reply:
I guess the part that I don't understand is what's different between going from `BECOMING_HOSTING` to `IS_PRIMARY_HOSTING` versus going from `VOLUNTEERING_HOSTING` to `IS_PRIMARY_HOSTING` that means we only care about the first case but not the second. In both cases the bucket starts off as not primary and then becomes primary.

Reply:
Ah, I understand now, thanks for the explanation!